Added ICAP CRC and ECC calculations and tests

herbertp opened the pull request:
Signed-off-by: Herbert Poetzl <herbert@13thfloor.at>

mithro commented:
Thanks @herbertp -- There are a couple of small formatting nits detected by travis - see https://travis-ci.org/SymbiFlow/prjxray/jobs/335069978

You should be able to fix this by running a `make format` and then updating your pull request.

Adding @kc8apf to review the actual code.

kc8apf commented on lib/CMakeLists.txt:
```diff
@@ -53,7 +53,9 @@ if (PRJXRAY_BUILD_TESTING)
 		xilinx/xc7series/part_test.cc
-		xilinx/xc7series/row_test.cc)
+		xilinx/xc7series/row_test.cc
+		xilinx/xc7series/crc_test.cc
```
Please keep the file list sorted alphabetically.

kc8apf commented on lib/include/prjxray/xilinx/xc7series/crc.h:
```diff
@@ -0,0 +1,32 @@
+// Extend the current CRC value with one address (5bit) and data (32bit) 
+// pair and return the newly computed CRC value
+
+uint32_t icap_crc(uint32_t addr, uint32_t data, uint32_t prev) {
```
I've been using CamelCase for method names.  Please elaborate on how addr and data are provided for a multi-word write.

kc8apf replied:
I follow https://google.github.io/styleguide/cppguide.html by habit which says CamelCase for both method and function names unless they are accessors or mutators.  I'm not set in stone on this so I'll leave it up to you to decide whether to change it.

kc8apf commented on lib/include/prjxray/xilinx/xc7series/crc.h:
```diff
@@ -0,0 +1,32 @@
+uint32_t icap_crc(uint32_t addr, uint32_t data, uint32_t prev) {
+	uint64_t val = ((uint64_t)addr << 32) | data;
+	uint64_t crc = prev;
+	uint64_t poly = 0x82F63B78L << 1;	// CRC-32C (Castagnoli)
```
Pull polynomial out into a file-scoped constexpr uint32_t with a descriptive name (e.g. kCrc32CastagnoliPolynomial).

kc8apf commented on lib/include/prjxray/xilinx/xc7series/crc.h:
```diff
@@ -0,0 +1,32 @@
+// pair and return the newly computed CRC value
+
+uint32_t icap_crc(uint32_t addr, uint32_t data, uint32_t prev) {
+	uint64_t val = ((uint64_t)addr << 32) | data;
```
Prefer static_cast<uint64_t>() over C-style casting.

kc8apf commented on lib/include/prjxray/xilinx/xc7series/crc.h:
```diff
@@ -0,0 +1,32 @@
+	uint64_t crc = prev;
+	uint64_t poly = 0x82F63B78L << 1;	// CRC-32C (Castagnoli)
+
+	for (int i = 0; i < 37; i++) {
```
Why 37?  Can this be explained by using a constexpr variable with a descriptive name instead?

kc8apf commented on lib/include/prjxray/xilinx/xc7series/ecc.h:
```diff
@@ -0,0 +1,50 @@
+namespace xc7series {
+
+// Extend the current ECC code with one data word (32 bit) at a given
+// index in the configuration packet and return the new ECC code.
```
I don't understand why index is defined relative to the configuration packet.  I thought the ECC was defined only in terms of a single frame.  Normal bitstreams use a single configuration packet to write the whole configuration.  Won't that wreak havoc on this method?

kc8apf commented on lib/include/prjxray/xilinx/xc7series/ecc.h:
```diff
@@ -0,0 +1,50 @@
+// index in the configuration packet and return the new ECC code.
+
+uint32_t icap_ecc(uint32_t idx, uint32_t data, uint32_t ecc) {
+	uint32_t off = idx * 32;
```
This looks like bit offset within the frame.  If that's what it is, please use a name that clearly says that.

kc8apf commented on lib/include/prjxray/xilinx/xc7series/ecc.h:
```diff
@@ -0,0 +1,50 @@
+// Extend the current ECC code with one data word (32 bit) at a given
+// index in the configuration packet and return the new ECC code.
+
+uint32_t icap_ecc(uint32_t idx, uint32_t data, uint32_t ecc) {
```
Is there an advantage to doing this calculation one word at a time instead of taking a collection or a pair of iterators to do the calculation over?

kc8apf commented on lib/xilinx/xc7series/crc_test.cc:
```diff
@@ -0,0 +1,13 @@
+
+TEST(IcapCrcTest, SimpleTests) {
+	EXPECT_EQ(xc7series::icap_crc(0, 0, 0), (uint32_t)0x0);
+	EXPECT_EQ(xc7series::icap_crc(~0, ~0, 0), 0xBF86D4DF);
```
What code path is this testing?  Why is ~0 a more relevant test compared to all other possible values?

herbertp commented:
On Tue, Jan 30, 2018 at 06:01:57PM +0000, Rick Altherr wrote:
> kc8apf requested changes on this pull request.

> Thanks! This will help considerably in writing valid frames.
> Most of my comments are related to helping others who are not
> familiar with the code understand what it is doing.

>> @@ -53,7 +53,9 @@ if (PRJXRAY_BUILD_TESTING)
>  		xilinx/xc7series/configuration_packetizer_test.cc
>  		xilinx/xc7series/global_clock_region_test.cc
>  		xilinx/xc7series/part_test.cc
> -		xilinx/xc7series/row_test.cc)
> +		xilinx/xc7series/row_test.cc
> +		xilinx/xc7series/crc_test.cc

> Please keep the file list sorted alphabetically.

It wasn't sorted, so I added it at the end.
The updated request has it sorted now.

>> @@ -0,0 +1,32 @@
> +#ifndef PRJXRAY_LIB_XILINX_XC7SERIES_CRC_H_
> +#define PRJXRAY_LIB_XILINX_XC7SERIES_CRC_H_
> +
> +#include <cstdint>
> +
> +namespace prjxray {
> +namespace xilinx {
> +namespace xc7series {
> +
> +// Extend the current CRC value with one address (5bit) and data (32bit) 
> +// pair and return the newly computed CRC value
> +
> +uint32_t icap_crc(uint32_t addr, uint32_t data, uint32_t prev) {

> I've been using CamelCase for method names. 

For now those are functions, should I still use CamelCase
for them like Sin() or ArcTan()?

> Please elaborate on how addr and data are provided for a
> multi-word write.

I've adjusted to comments to explain that addr is the 
register address the data is written to.

>> +#ifndef PRJXRAY_LIB_XILINX_XC7SERIES_CRC_H_
> +#define PRJXRAY_LIB_XILINX_XC7SERIES_CRC_H_
> +
> +#include <cstdint>
> +
> +namespace prjxray {
> +namespace xilinx {
> +namespace xc7series {
> +
> +// Extend the current CRC value with one address (5bit) and data (32bit) 
> +// pair and return the newly computed CRC value
> +
> +uint32_t icap_crc(uint32_t addr, uint32_t data, uint32_t prev) {
> +	uint64_t val = ((uint64_t)addr << 32) | data;
> +	uint64_t crc = prev;
> +	uint64_t poly = 0x82F63B78L << 1;	// CRC-32C (Castagnoli)

> Pull polynomial out into a file-scoped constexpr uint32_t with
> a descriptive name (e.g. kCrc32CastagnoliPolynomial).

Done.

>> @@ -0,0 +1,32 @@
> +#ifndef PRJXRAY_LIB_XILINX_XC7SERIES_CRC_H_
> +#define PRJXRAY_LIB_XILINX_XC7SERIES_CRC_H_
> +
> +#include <cstdint>
> +
> +namespace prjxray {
> +namespace xilinx {
> +namespace xc7series {
> +
> +// Extend the current CRC value with one address (5bit) and data (32bit) 
> +// pair and return the newly computed CRC value
> +
> +uint32_t icap_crc(uint32_t addr, uint32_t data, uint32_t prev) {
> +	uint64_t val = ((uint64_t)addr << 32) | data;

> Prefer static_cast<uint64_t>() over C-style casting.

Done

>> +
> +#include <cstdint>
> +
> +namespace prjxray {
> +namespace xilinx {
> +namespace xc7series {
> +
> +// Extend the current CRC value with one address (5bit) and data (32bit) 
> +// pair and return the newly computed CRC value
> +
> +uint32_t icap_crc(uint32_t addr, uint32_t data, uint32_t prev) {
> +	uint64_t val = ((uint64_t)addr << 32) | data;
> +	uint64_t crc = prev;
> +	uint64_t poly = 0x82F63B78L << 1;	// CRC-32C (Castagnoli)
> +
> +	for (int i = 0; i < 37; i++) {

> Why 37? Can this be explained by using a constexpr variable
> with a descriptive name instead?

Done.

>> @@ -0,0 +1,50 @@
> +#ifndef PRJXRAY_LIB_XILINX_XC7SERIES_ECC_H_
> +#define PRJXRAY_LIB_XILINX_XC7SERIES_ECC_H_
> +
> +#include <cstdint>
> +
> +namespace prjxray {
> +namespace xilinx {
> +namespace xc7series {
> +
> +// Extend the current ECC code with one data word (32 bit) at a given
> +// index in the configuration packet and return the new ECC code.

> I don't understand why index is defined relative to the
> configuration packet. I thought the ECC was defined only
> in terms of a single frame. Normal bitstreams use a single
> configuration packet to write the whole configuration. 
> Won't that wreak havoc on this method?

My fault, adjusted the comment section accordingly.

>> @@ -0,0 +1,50 @@
> +#ifndef PRJXRAY_LIB_XILINX_XC7SERIES_ECC_H_
> +#define PRJXRAY_LIB_XILINX_XC7SERIES_ECC_H_
> +
> +#include <cstdint>
> +
> +namespace prjxray {
> +namespace xilinx {
> +namespace xc7series {
> +
> +// Extend the current ECC code with one data word (32 bit) at a given
> +// index in the configuration packet and return the new ECC code.
> +
> +uint32_t icap_ecc(uint32_t idx, uint32_t data, uint32_t ecc) {
> +	uint32_t off = idx * 32;

> This looks like bit offset within the frame.  
> If that's what it is, please use a name that clearly says that.

Added a comment and renamed it to avoid confusion.

>> @@ -0,0 +1,50 @@
> +#ifndef PRJXRAY_LIB_XILINX_XC7SERIES_ECC_H_
> +#define PRJXRAY_LIB_XILINX_XC7SERIES_ECC_H_
> +
> +#include <cstdint>
> +
> +namespace prjxray {
> +namespace xilinx {
> +namespace xc7series {
> +
> +// Extend the current ECC code with one data word (32 bit) at a given
> +// index in the configuration packet and return the new ECC code.
> +
> +uint32_t icap_ecc(uint32_t idx, uint32_t data, uint32_t ecc) {

> Is there an advantage to doing this calculation one word at a
> time instead of taking a collection or a pair of iterators to
> do the calculation over?

Same as with the CRC, it can be done 'while' you are 
scanning data and you don't need to go over the data
twice.

Can still be wrapped for generating the data for writes.

>> @@ -0,0 +1,13 @@
> +#include <prjxray/xilinx/xc7series/crc.h>
> +
> +#include <gtest/gtest.h>
> +
> +namespace xc7series = prjxray::xilinx::xc7series;
> +
> +TEST(IcapCrcTest, SimpleTests) {
> +	EXPECT_EQ(xc7series::icap_crc(0, 0, 0), (uint32_t)0x0);
> +	EXPECT_EQ(xc7series::icap_crc(~0, ~0, 0), 0xBF86D4DF);

> What code path is this testing? 
> Why is ~0 a more relevant test compared to all other possible
> values?

Added comments to all of the test cases.

> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/SymbiFlow/prjxray/pull/70#pullrequestreview-92656664

kc8apf commented on lib/include/prjxray/xilinx/xc7series/crc.h:
```diff
@@ -3,19 +3,25 @@
+	uint64_t val = (static_cast<uint64_t>(addr) << 32) | data;
 	uint64_t crc = prev;
-	uint64_t poly = 0x82F63B78L << 1;	// CRC-32C (Castagnoli)
+	constexpr int kFivePlusThrityTwo = 37;
```
Thirty is misspelled.  Further, that doesn't help me understand why it is 5+32.  kAddressBitWidth and kDataBitWidth with (kAddressBitWidth + kDataBitWidth) in the loop condition makes a lot more sense to me.

kc8apf commented on lib/include/prjxray/xilinx/xc7series/ecc.h:
```diff
@@ -8,36 +8,35 @@ namespace xilinx {
-		off += 0x1340;
-	else				// avoid lower
-		off += 0x1320;
+	if (idx > 0x25)  // avoid 0x800
```
Not required in this PR but I'd like to see a writeup on the ECC calculation including why these special cases exist.

herbertp commented:
On Wed, Jan 31, 2018 at 06:24:22PM +0000, Rick Altherr wrote:
> kc8apf requested changes on this pull request.

>> @@ -0,0 +1,32 @@
> +#ifndef PRJXRAY_LIB_XILINX_XC7SERIES_CRC_H_
> +#define PRJXRAY_LIB_XILINX_XC7SERIES_CRC_H_
> +
> +#include <cstdint>
> +
> +namespace prjxray {
> +namespace xilinx {
> +namespace xc7series {
> +
> +// Extend the current CRC value with one address (5bit) and data (32bit) 
> +// pair and return the newly computed CRC value
> +
> +uint32_t icap_crc(uint32_t addr, uint32_t data, uint32_t prev) {

> I follow https://google.github.io/styleguide/cppguide.html by
> habit which says CamelCase for both method and function names
> unless they are accessors or mutators. I'm not set in stone on
> this so I'll leave it up to you to decide whether to change it.

Okay, I'll stick with the lower case function names then
(similar to the already present bit_* operations).

>>  	uint64_t crc = prev;
> -	uint64_t poly = 0x82F63B78L << 1;	// CRC-32C (Castagnoli)
> +	constexpr int kFivePlusThrityTwo = 37;

> Thirty is misspelled. Further, that doesn't help me understand
> why it is 5+32. kAddressBitWidth and kDataBitWidth with
> (kAddressBitWidth + kDataBitWidth) in the loop condition makes
> a lot more sense to me.

Done.

> -	if (idx > 0x25)			// avoid 0x800
> -		off += 0x1360;
> -	else if (idx > 0x6)		// avoid 0x400
> -		off += 0x1340;
> -	else				// avoid lower
> -		off += 0x1320;
> +	if (idx > 0x25)  // avoid 0x800

> Not required in this PR but I'd like to see a writeup on the
> ECC calculation including why these special cases exist.

ECC is about correcting bit errors, in this case single bit
errors, so the ECC code needs to record which bit is set.
The simplest way to do that is to number the bits. In our
case there are 101 words of 32 bits so a total of 3232 bits.
Now obviously if you start to count at '0' you have a problem
because you wouldn't be able to tell 'bit 0' from 'no bit',
and starting at '1' would have a similar problem when a single
bit changes (could be either the first bit or the '1' in the
ECC code). Similar problems arise when you have single bits
set in the ECC code, i.e. powers of two, so you naturally want
to avoid those. A good approach is to start counting at the
end (highest value) and count down, i.e. bit 3232 is 0xFFF,
3231 is 0xFFE and so on. You still have to avoid 0x800 and
0x400 on your way down.

Best,
Herbert

> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/SymbiFlow/prjxray/pull/70#pullrequestreview-93028434

kc8apf commented:
Thank you for the explanation on ECC.  I'll move that to a Wiki page or other doc soon.
